controller/tests/models/domain/imagefilter_test.py

```diff
@@ -0,0 +1,191 @@
+    # tags in the resulting range, and then check to make sure the middle one
+    # is always present, which it should be.
+
+    now = datetime.now(tz=UTC)
+
+    last_week = now - timedelta(days=7)
+    two_weeks_ago = now - timedelta(days=14)
+    three_weeks_ago = now - timedelta(days=21)
+
+    yesterday = now - timedelta(days=1)
+    day_before_yesterday = now - timedelta(days=2)
+    three_days_ago = now - timedelta(days=3)
+
+    tag_this_week = to_weekly_tag(now)
+    tag_last_week = to_weekly_tag(last_week)
+    tag_two_weeks_ago = to_weekly_tag(two_weeks_ago)
+    tag_three_weeks_ago = to_weekly_tag(three_weeks_ago)
+
+    tag_today = to_daily_tag(now)
+    tag_yesterday = to_daily_tag(yesterday)
+    tag_day_before_yesterday = to_daily_tag(day_before_yesterday)
+    tag_three_days_ago = to_daily_tag(three_days_ago)
+
+    exp_tag_this_week = f"exp_{tag_this_week}"
+    exp_tag_last_week = f"exp_{tag_last_week}"
+    exp_tag_two_weeks_ago = f"exp_{tag_two_weeks_ago}"
+
+    tags = [
+        "recommended",
+        "r28_0_1",
+        "r28_0_0",
+        "r27_0_0",
+        "r26_0_0",
+        tag_this_week,
+        tag_last_week,
+        tag_two_weeks_ago,
+        tag_three_weeks_ago,
+        tag_today,
+        tag_yesterday,
+        tag_day_before_yesterday,
+        tag_three_days_ago,
+        "r28_0_0_rc1",
+        "r27_0_0_rc1",
+        exp_tag_this_week,
+        exp_tag_last_week,
+        exp_tag_two_weeks_ago,
+        "unknown",
+    ]
+
+    shuffled_tags = list(tags)
+    SystemRandom().shuffle(shuffled_tags)
+
+    recommended = {"recommended"}
+    collection = RSPImageTagCollection.from_tag_names(
+        shuffled_tags, recommended
+    )
+
+    assert len(list(collection.all_tags())) == len(tags)
+
+    # Create image policy
+
+    policy = RSPImageFilterPolicy(
+        release=ImageFilterPolicy(
+            # We should get three
+            cutoff_version=Version(major=27, minor=0, patch=0)
+        ),
+        weekly=ImageFilterPolicy(
+            # We should get one, two, or three
+            age=timedelta(weeks=2)
+        ),
+        daily=ImageFilterPolicy(
+            # We should get one, two, or three from the age policy,
+            # but definitely not four: number will never be the filter
+            age=timedelta(days=2),
+            number=4,
+        ),
+        release_candidate=ImageFilterPolicy(
+            # We should only get one, not two: number will be the filter
+            cutoff_version=Version(major=25, minor=3, patch=1),
+            number=1,
+        ),
+        experimental=ImageFilterPolicy(
+            # We should get one, two or three.  For
+            # dated releases we use calver
+            cutoff_version=Version(
+                major=last_week.isocalendar().year,
+                minor=last_week.isocalendar().week,
+                patch=0,
+            )
+        ),
+    )
+
+    filtered = collection.apply_policy(policy)
```

Comment:
`apply_policy` (or `filter` per above) should take a time basis for the filtering, so you can pass in `now`. That will fix your problems with having a variable number of tags in some category and simplify a bunch of the rest of the test suite, since then you can probably just check the result of `all_tags` against an expected list instead of checking each class separately.